aws/resource_aws_s3_bucket.go

```diff
@@ -1348,6 +1448,73 @@ func resourceAwsS3BucketPolicyUpdate(s3conn *s3.S3, d *schema.ResourceData) erro
 	return nil
 }
 
+func resourceAwsS3BucketGrantsUpdate(s3conn *s3.S3, d *schema.ResourceData) error {
+	bucket := d.Get("bucket").(string)
+	rawGrants := d.Get("grant").([]interface{})
+
+	if len(rawGrants) == 0 {
+		log.Printf("[DEBUG] S3 bucket: %s, Grants fallback to canned ACL", bucket)
+		if err := resourceAwsS3BucketAclUpdate(s3conn, d); err != nil {
+			return fmt.Errorf("Error fallback to canned ACL, %s", err)
+		}
+	} else {
+		apResponse, err := retryOnAwsCode("NoSuchBucket", func() (interface{}, error) {
+			return s3conn.GetBucketAcl(&s3.GetBucketAclInput{
+				Bucket: aws.String(d.Id()),
+			})
+		})
+
+		ap := apResponse.(*s3.GetBucketAclOutput)
+		if err != nil {
+			return err
```

Comment:
We should return error context here for operators and code maintainers

```suggestion
			return fmt.Errorf("error getting S3 Bucket (%s) ACL: %s", d.Id(), err)
```

Reply:
Done